### ISSUE TITLE
Drop 3DNow! Simd code

### ISSUE THREAD

**Issue by tyabus** [dhewm/dhewm3#443]
I don't think we need this anymore, even linux kernel is dropping it and AMD dosent seem to support them either, so what's the point of keeping it?

**Comment by DanielGibson** [comment]
What's the point in removing it? I never even touched that code.

Linux kernel dropping it is irrelevant, as it's Windows+MSVC-only anyway

**Comment by paulwratt** [comment]
> I don't think we need this anymore, even linux kernel is dropping it and AMD dosent seem to support them either, so what's the point of keeping it?

Then it wont run on 3DNow equipped hardware. Well it will, but _without_ the resulting 3DNow speed enhancements.

**Comment by SingingBush** [comment]
> What's the point in removing it?

3DNow was deprecated over a decade ago. Supporting hardware that far back as well as current systems isn't likely to be maintainable so removing it all together probably makes sense.

**Comment by paulwratt** [comment]
_dont_ take it out.

if you want to remove it from code, then record it as a patch in the repo.

> 3DNow was deprecated over a decade ago.

_not_ a good enough reason, especially if you are a quake playing owner of a worrking 3DNow system.

**Comment by Yamagi** [comment]
It ain't broken, so don't fix it :) And yes, on old Athlon64 and Athlon X2 systems 3DNow makes a difference. Only a few frames per second, however in some situations - especially the big outdoor scenes in RoE - they can decide between unplayable stuttering or slow but somewhat okay. My shiny dual-socket Opteron X2 265 with 1.8 GHz was a good example for that.

**Comment by Kasreyn** [comment]
I don't think 3DNow! support is required to be in the Linux kernel for applications to take advantage of it.
What I mean is, dhewm3 could still be compiled to use 3DNow! resulting in a performance benefit.

Devuan 6 fully supports older 32bit x86 like the Athlon XP, and I can tell you,
it runs fast and feels responsive, no problemo!

It will never be my daily driver (as it once was) all I'm saying is: these old "war ships" exist.
And wouldn't you agree that it's great, these "older computers" can still use the latest versions of open source "core" tools?

**Comment by ClickClick5** [comment]
Well, over the better part of a year (starting with 1.5.3), I took it upon myself as a fun project to utterly rebuild the SIMD setup.
Challenge one, compile the game as native 64 bit...which does not ingest any __asm, thus the 64bit windows compile always defaults to simple scaler math. After a lengthy uphill battle, and finally caving in and having Claude help track down some Transform issues, I have a almost healthy, FULL SSE4.1 only, 64bit native compile. With cpu cylcles drastically reduced, the game's arithmetic performs about 4-6x faster than before. 

I know this can't be put into the project as canon due to AI use (just gosh darn-it...single handedly rebuilding the entire mathematical code path was daunting), but it's worth to know it IS doable by hand over all.

In short, only SSE4.1 is run, native x64 for windows. Thus any CPU without SSE4.1 will be met with an immediate runtime crash starting the game. 

My original goal was to push to AVX/AVX2, but quickly realized that the change of the code would have been far more involved than staying in the SSE field. And with Doom 3's logic, AVX is honestly not going to improve anything.

EDIT: The fallback scaler, MMX through SSE3 (Simd_SSE3.cpp) files and code have been removed entirely. This is a high compile for modern era hardware and Windows 10/11 x64.

**Comment by nbohr1more** [comment]
> Well, over the better part of a year (starting with 1.5.3), I took it upon myself as a fun project to utterly rebuild the SIMD setup. Challenge one, compile the game as native 64 bit...which does not ingest any __asm, thus the 64bit windows compile always defaults to simple scaler math. After a lengthy uphill battle, and finally caving in and having Claude help track down some Transform issues, I have a almost healthy, FULL SSE4.1 only, 64bit native compile. With cpu cylcles drastically reduced, the game's arithmetic performs about 4-6x faster than before.
> 
> I know this can't be put into the project as canon due to AI use (just gosh darn-it...single handedly rebuilding the entire mathematical code path was daunting), but it's worth to know it IS doable by hand over all.
> 
> In short, only SSE4.1 is run, native x64 for windows. Thus any CPU without SSE4.1 will be met with an immediate runtime crash starting the game.
> 
> My original goal was to push to AVX/AVX2, but quickly realized that the change of the code would have been far more involved than staying in the SSE field. And with Doom 3's logic, AVX is honestly not going to improve anything.
> 
> EDIT: The fallback scaler, MMX through SSE3 (Simd_SSE3.cpp) files and code have been removed entirely. This is a high compile for modern era hardware and Windows 10/11 x64.

The Dark Mod has native AVX / AVX2 support so that can be ported eventually...

**Comment by DanielGibson** [comment]
Yeah, and if running on a CPU without SSE4.1 crashes the game you (or your chatbot) have done something fundamentally wrong, the whole point of these idSIMD_* classes is having multiple implementations for different supported extensions, so one that suits your CPU can be used.

I may look into all this at some point.. I'd probably try to have a working backends (that use intrinsics instead of inline ASM) for:
* SSE2, because every x86_64/x64/amd64 CPU, i.e. every remotely relevant x86 CPU, supports it
* AVX or AVX2 (which may include (S)SSE3 or 4 or whatever if all CPUs supporting AVX support those as well) or something for newer CPUs
* Ideally also a backend for ARM NEON

Maybe not for all functions, but for the ones that are most relevant (which has to be determined with profiling), I can imagine that many work well enough with the C++ fallback code when using halfway modern CPUs and compilers.

TDM has some additional SIMD functions that dhewm3 doesn't have, porting those (and the places calling them) specifically probably also makes sense.

<br>

But right now I'm just glad I've released 1.5.5 and am looking forward to spending time on other projects before coming back to dhewm3 eventually.

<br>

Of course if anyone wants to do the suggested changes (themselves, not with AI) that would be welcome :)

**Comment by ClickClick5** [comment]
It crashes because I removed the fallbacks. It's taken the better part of two years to just get to this point. In doing this, I'm fully aware it breaks all the compatibility of other systems, but the point of this was to...essentially create a happy 64bit compile for windows, as if the game was released today. This was a project of speed and modernization over compatibility.

I would love to do these changes without AI! But that would not be ready before game version 2.7.9 😆 

This was honestly more of a headache then intended: removing all __asm and converting to intrinsics for compiling.

Also, was not aware of the dark mod. And Daniel, my man, take a rest, you deserve it!!!

**Comment by DanielGibson** [comment]
> I would love to do these changes without AI! But that would not be ready before game version 2.7.9

I'm sure with practice it gets easier ;)

BTW, D3BFG has an intrinsics version for SSE1 at least: https://github.com/klaussilveira/chocolate-doom3-bfg/blob/master/neo/idlib/math/Simd_SSE.cpp

**Comment by ClickClick5** [comment]
The scaler fallback math was the last bits I removed, I -could- add that back in bahaha!

But yes, I can go back through and commit other smaller changes.

For example, MixedSoundToSamples in Simd_SSE2.cpp has a note from iD about the cache pollution that takes place. Redone, it saves about 2-4 cpu cycles per every 8 samples, bypassing the L1/L2 cache.

//SSE4 rewrite for ID's "// FIXME: should not use movlps/movhps to move integer data" in Simd_SSE2.cpp, line 866
void VPCALL idSIMD_SSE4::MixedSoundToSamples( short *samples, const float *mixBuffer, const int numSamples ) {
    assert( ( numSamples % MIXBUFFER_SAMPLES ) == 0 );

    for ( int i = 0; i < numSamples; i += 8 ) {
        __m128 f0 = _mm_load_ps( mixBuffer + i + 0 );
        __m128 f1 = _mm_load_ps( mixBuffer + i + 4 );
        _mm_prefetch( (const char *)(mixBuffer + i + 32), _MM_HINT_NTA );
        __m128i i0 = _mm_cvtps_epi32( f0 );
        __m128i i1 = _mm_cvtps_epi32( f1 );
        __m128i packed = _mm_packs_epi32( i0, i1 );  // Love me that room [-32768, 32767]
        _mm_stream_si128( (__m128i *)(samples + i), packed );
    }
    _mm_sfence();
}

Alas, again, this was a bugger of a task. It's still not 100% and small lighting bugs are being noticed, but I'm still cutting through it all.

**Comment by DanielGibson** [comment]
And how big are the performance benefits?

**Comment by ClickClick5** [comment]
Not much here, just better code cleanup. The old SSE2 stored 16 bytes in 4 stores (2 stores, 8 bytes (half of an XMM register)). This version stores the entire call in one 16 byte store, with no cache allocation.

16 bytes at a time, a 50% memory bandwidth increase. Is this particular method noticeable for the player on modern hardware? Nope. But fixes a bad implementation, and modernizes the structure.

There are many methods that saw ~2-8% increases in speed, some saw none (but the code is now an intrinsic). Some methods saw a massive 50%+ increase, but these used a little bit of AI to solve.

**Comment by ClickClick5** [comment]
Well as to actually help out here:

In Simd_SSE2.cpp, delete lines 866 through 869, then add:

movntdq	[esi-4 * 4*  2], xmm4
movntdq	[esi-2 * 4 * 2], xmm6

On line 871, add: sfence

This should fix the L1 cpu cache bug that id never got around to fixing. It frees up 256 bytes of L1 cache being wasted on every audio call.

**Comment by DanielGibson** [comment]
That function (`MixedSoundToSamples()`) isn't used anymore.
It was only used by the old sound backend, not by the OpenAL backend.

It's only called from some test function

**Comment by ClickClick5** [comment]
Mmm, well after passing through the MMX-SSE3 asm, there is not much to optimize apart from intrinsic conversions. So there we have it lol
